src/Approximators/RangeApproximators/rangefinder.jl

```
@@ -0,0 +1,150 @@
+    be orthogonalizing each matrix in what is known as the random subspace iteration.
+
+# Fields
+- `compressor::Compressor`, the technique that will compress the matrix from the right.
+- `power_its::Int64`, the number of power iterations that should be performed.
+- `rand_subspace::Bool`, a boolean indicating whether the power its should be performed 
+    with orthogonalization.
+"""
+mutable struct RangeFinder <: RangeApproximator
+    compressor::Compressor
+    power_its::Int64
+    rand_subspace::Bool
+    function RangeFinder(compressor, power_its, rand_subspace)
+        if power_its < 0
+            throw(ArgumentError("Field `power_its` must be non-negative."))
+        end
+        
+        return new(compressor, power_its, rand_subspace)
+    end
+
+end
+
+RangeFinder(;
+    compressor = SparseSign(), 
+    rand_subspace = false, 
+    power_its = 1
+   ) = RangeFinder(compressor, rand_subspace, power_its)
+
+"""
+    RangeFinderRecipe
+
+A struct that contains the preallocated memory and completed compressor to form a
+    RangeFinder approximation to the matrix ``A``.
+
+# Fields
+- `compressor::CompressorRecipe`, the compressor to be applied from the right to ``A``.
+- `power_its::Int64`, the number of power iterations that should be performed.
+- `rand_subspace::Bool`, a boolean indicating whether the power its should be performed 
+    with orthogonalization.
+- `range::AbstractMatrix`, the orthogonal matrix that approximates the range of ``A``.
+"""
+mutable struct RangeFinderRecipe <: RangeApproximatorRecipe
+    n_rows::Int64
+    n_cols::Int64
+    compressor::CompressorRecipe
+    power_its::Int64
+    rand_subspace::Bool
+    range::AbstractMatrix
+end
+
+function complete_approximator(approx::RangeFinder, A::AbstractMatrix)
+    type = eltype(A)
+    # You need to make sure you orient the compressor in the correct direction
+    if typeof(approx.compressor.cardinality) <: Left
+        approx.compressor.cardinality = Right()
+    end
```

What is going on here? It is better to throw an error here no?

---

In all cases, we want the compressor to be applied from the left. To me it's therefore better to change any right compressor to the correct left compressor, and possibly return a warning. I don't think returning an error is really worthwhile in this instance.

---

I do not think this is safe behavior. I think the error can tell them they want to pass a `Right()` cardinality compressor or modify it themselves. If your function is changing a data structure without you asking for the change, that is not a great idea. You can throw a warning and at the end, change the cardinality back to `Left()`.

---

This is what I have done.